chess/ai/minimax.py

```diff
@@ -0,0 +1,171 @@
+
+            if m is not None:
+                if m[0] > v[0]:
+                    v[0] = m[0]
+
+            v[1] = source
+            v[2] = movement[1]
+
+            v_list.append(v)
+
+            if v[0] <= alfa:
+                if self.jogada_aleatoria(v_list):
+                    if len(v_list) != 0:
+                        return v_list[randint(0, len(v_list) - 1)]
+                else:
+                    return min(v_list, key=lambda v: v[0])
+
+            beta = min(v[0], beta)
+
+        if self.jogada_aleatoria(v_list):
+            if len(v_list) != 0:
+                return v_list[randint(0, len(v_list) - 1)]
+        else:
+            return min(v_list, key=lambda v: v[0])
+
+    # Maximiza a jogada do PC (peça preta)
+    def max_value(self, state, alfa, beta, depth):
+        # o algoritmo para quando chega na profundidade especificada e
+        # retorna o valor da ultima jogada e para a recursão
+        if self.cutoff_test(depth):
+            return [state.value, None, None]
+
+        if is_checkmate_for_player(self.state.game, Player.WHITE):
+            return [self.VICTORY, None]
+
+        v_list = []
+
+        vm = self.valid_movements(state, BLACK_PIECES)
+        for movement in vm:
+            v = [float("-inf"), None, None]
+            source = movement[0]
+            maxstate = deepcopy(state)
+            # faz a jogada
+            move(maxstate.game, source, movement[1])
+            # valor da jogada
+            v[0] = score_board(maxstate.game.board)
+
+            # associa o valor da jogada com o estado atual
+            maxstate.value = v[0]
+
+            # momento em que pega a jogada do adversário
+            m = self.min_value(maxstate, alfa, beta, depth - 1)
+
+            if m is not None:
+                if m[0] < v[0]:
+                    v[0] = m[0]
+
+            v[1] = source
+            v[2] = movement[1]
+
+            v_list.append(v)
+
+            if v[0] >= beta:
+                if self.jogada_aleatoria(v_list):
+                    if len(v_list) != 0:
+                        return v_list[randint(0, len(v_list) - 1)]
+                else:
+                    return max(v_list, key=lambda v: v[0])
+
+            alfa = max(v[0], alfa)
+
+        if self.jogada_aleatoria(v_list):
+            if len(v_list) != 0:
+                return v_list[randint(0, len(v_list) - 1)]
+        else:
+            return max(v_list, key=lambda v: v[0])
+
+    def valid_movements(self, state, piece_type):
+        destinations_list = []
+
+        for row in range(8):
+            for column in range(8):
+                piece = state.game.board[row][column]
+                if piece in piece_type:
+
+                    src = Coordinate(row, column)
+
+                    for dest in destinations(state.game, src):
+
+                        destinations_list.append((src, dest))
+
+        return destinations_list
+
+    def jogada_aleatoria(self, l):
```

Comment:
Vc quer fazer isso pra quando o score for o máximo? Pois vão ter casos onde teremos mais de uma jogada com o mesmo valor e esse valor pode ser o máximo dentre aquelas jogadas, e não a melhor jogada possível, entende?